l1-contracts/src/core/Rollup.sol

```diff
@@ -71,12 +79,26 @@ contract Rollup is IRollup {
       revert Errors.Rollup__UnavailableTxs(header.contentCommitment.txsEffectsHash);
     }
 
-    bytes32[] memory publicInputs = new bytes32[](1);
-    publicInputs[0] = _computePublicInputHash(_header, _archive);
+    bytes32[] memory publicInputs = new bytes32[](40);
+    publicInputs[0] = _archive;
+    // this is the _next_ available leaf in the archive tree
+    // normally this should be equal to the block number (since leaves are 0-indexed and blocks 1-indexed)
+    // but in yarn-project/merkle-tree/src/new_tree.ts we prefill the tree so that block N is in leaf N
+    publicInputs[1] = bytes32(header.globalVariables.blockNumber + 1);
+
+    bytes32[22] memory headerFields = HeaderLib.toFields(header);
+    for (uint256 i = 0; i < headerFields.length; i++) {
+      publicInputs[i + 2] = headerFields[i];
+    }
+
+    // the block proof is recursive, which means it comes with an aggregation object
+    // this snippet copies it into the public inputs needed for verification
+    // it also guards against empty _aggregationObject used with mocked proofs
+    for (uint256 i = 0; i < 16 && i * 32 < _aggregationObject.length; i++) {
+      publicInputs[i + 24] = bytes32(_aggregationObject[i * 32:(i + 1) * 32]);
+    }
```

Comment:
Food for thought:
```solidity
uint256 aggregationLength = _aggregationObject.length / 32;
for (uint256 i = 0; i < 16 && i < aggregationLength; i++) {
    bytes32 part;
    assembly {
        part := mload(add(_aggregationObject, add(32, mul(i, 32))))
    }
    publicInputs[i + 24] = part;
}
```

Reply:
Thanks for the idea! I ended up taking this and refactoring it to use `calldataload`.